napari/viewer.py

```diff
@@ -30,6 +31,17 @@ class Viewer(ViewerModel):
     # Create private variable for window
     _window: 'Window'
 
+    _instances: ClassVar[list] = []
+
+    @classmethod
+    def active_instance(self):
+        all_refs = [v() for v in self._instances]
+        not_none = [v for v in all_refs if v is not None]
+        actives = [v for v in not_none if v.window._qt_window.hasFocus()]
```

Comment:
Right now this predicate does not work.

Reply:
`hasFocus` is never going to work well if we want ipython to be a "first class" use case... since as soon as you go back to ipython, _no_ windows will have focus.  
Also, since what we generally care about is the last used "window"... I put this instance-saving logic on QtMainWindow.  If what you need is a reference to the viewer itself, it's currently a bit ugly (but could be made easier.

try:
```python
napari._qt.qt_main_window._QtMainWindow.current().qt_viewer.viewer
```

Reply:
I also left all of this in the private API for now... it's not immediately clear to me that the Viewer method needs such a method, since the majority of our need for the "current viewer" is internal?

Reply:
I'd also prefer this to be private for now. I do find it odd from an API design perspective that the viewer can get an active viewer that is different from itself - that's not true for our layers and it makes things easier to think about (I think). I wonder if one day we might want to have something like

```python
application # a singleton for which there is only ever one
application.viewers # a set of viewers
application.viewers.active # the active viewer
```

If you're just using a single viewer then you never have to think about or worry about `application` or other `viewers`, it's all behind the scenes.

Note we don't need to do this now, but we might want to keep the new stuff private if we do